src/coreclr/src/jit/assertionprop.cpp

```diff
@@ -2163,22 +2163,20 @@ void Compiler::optAssertionGen(GenTree* tree)
             break;
 
         case GT_CALL:
+        {
             // A virtual call can create a non-null assertion. We transform some virtual calls into non-virtual calls
             // with a GTF_CALL_NULLCHECK flag set.
-            if ((tree->gtFlags & GTF_CALL_NULLCHECK) || tree->AsCall()->IsVirtual())
+            // Ignore tail calls because they have 'this` pointer in the regular arg list and an implicit null check.
+            GenTreeCall* const call = tree->AsCall();
+            if (call->NeedsNullCheck() || (call->IsVirtual() && !call->IsTailCall()))
```

Comment:
In `fgMorphTailCallViaJitHelper` we clear `NeedsNullCheck`, should not we also clear `IsVirtual`?
Then this condition will be just `if (call->NeedsNullCheck() || call->IsVirtual())`

Reply:
@jakobbotsch could you please advise me here?

Reply:
Sorry I don't know. `fgMorphTailCallViaJitHelper` is the old (from before #341) tailcall mechanism used only on x86. This mechanism might be used directly for virtual calls (or VSDs), but I don't remember how the address is computed in those cases.